### PR TITLE
*: handle signed/unsigned in the partition pruning

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve?

Problem Summary:

The old `lessThanData` use int64 for data representation.
However, the partition by range less than value can be signed or unsigned.
If it's unsigned, it may be out of the int64 range.

### What is changed and how it works?

What's Changed:

- move `makeLessThanData` processing to the table creating to reduce allocation
- handle signed/unsigned in compare
- reuse the session's parser in `expression.ParseSimpleExprsWithNames`

How it Works:

Compare function now use an unsigned flag to handle the unsigned case.

### Check List 

Tests 

- Unit test

### CODE REVIEW

**Comment by tiancaiamao** [comment]
PTAL @lysu @imtbkcat

**Comment by tiancaiamao** [comment]
Some fail cases are range columns partition, we have to merge this one first
https://github.com/pingcap/tidb/pull/15169

**Comment by tiancaiamao** [comment]
PTAL @lysu @imtbkcat

File: planner/core/rule_partition_processor.go [hunk 1]
```diff
@@ -328,37 +339,24 @@ func (s *partitionProcessor) pruneRangePartition(ds *DataSource, pi *model.Parti
...
-		// TODO: Store LessThanData in the partitionExpr, avoid allocating here.
-		lessThan, err := makeLessThanData(pi)
+		partExpr, err := ds.table.(partitionTable).PartitionExpr()
 		if err != nil {
 			return nil, err
 		}
-		pruner := rangePruner{lessThan, col, fn}
+		pruner := rangePruner{
+			lessThan: lessThanDataInt{
+				data:     partExpr.ForRangePruning.LessThan,
```
Comment (imtbkcat) [inline planner/core/rule_partition_processor.go]: Using copy here is better? Copy could avoid race here.

Comment (tiancaiamao) [reply planner/core/rule_partition_processor.go]: It's read-only data... free from data race, so the copy is unnecessary.

The `LessThan` is different from `expression.Expression` or `ast.StmtNode`
The [visit function](https://github.com/pingcap/parser/blob/master/ast/ast.go#L154) modify the node like [here](https://github.com/pingcap/parser/blob/cd90cc2a7d870118b3319bf50564a6d55613c145/ast/expressions.go#L201):

```
node.XX = XX.Accept(Visitor)
``` 

And the expression is also modified during use (ugly code).

**Comment by tiancaiamao** [comment]
PTAL @lysu

File: expression/simple_rewriter.go [hunk 2]
```diff
@@ -41,10 +43,20 @@ type simpleRewriter struct {
...
 	exprStr = "select " + exprStr
-	stmts, warns, err := parser.New().Parse(exprStr, "", "")
-	for _, warn := range warns {
-		ctx.GetSessionVars().StmtCtx.AppendWarning(util.SyntaxWarn(warn))
+	var stmts []ast.StmtNode
+	var err error
+	if p, ok := ctx.(interface {
+		ParseSQL(context.Context, string, string, string) ([]ast.StmtNode, []error, error)
+	}); ok {
+		stmts, _, err = p.ParseSQL(context.Background(), exprStr, "", "")
```
Comment (lysu) [inline expression/simple_rewriter.go]: why reusable parser didn't run `AppendWarning` if it meet warn~?

File: expression/simple_rewriter.go [hunk 3]
```diff
@@ -102,9 +114,18 @@ func ParseSimpleExprsWithSchema(ctx sessionctx.Context, exprStr string, schema *
...
```
Comment (lysu) [inline expression/simple_rewriter.go]: ditto

Comment (tiancaiamao) [reply expression/simple_rewriter.go]: Done

**Pull request merged**